homeassistant/components/telegram_bot/__init__.py

```diff
@@ -51,44 +198,296 @@ def async_setup_platform(p_type, p_config=None, discovery_info=None):
+        self.bot = Bot(token=api_key)
+        self.hass = hass
+
+    def _get_msg_ids(self, msg_data, chat_id):
+        """Get the message id to edit.
+
+        This can be one of (message_id, inline_message_id) from a msg dict,
+        returning a tuple.
+        **You can use 'last' as message_id** to edit
+        the last sended message in the chat_id.
+        """
+        message_id = inline_message_id = None
+        if ATTR_MESSAGEID in msg_data:
+            message_id = msg_data[ATTR_MESSAGEID]
+            if (isinstance(message_id, str) and (message_id == 'last') and
+                    (self._last_message_id[chat_id] is not None)):
+                message_id = self._last_message_id[chat_id]
+        else:
+            inline_message_id = msg_data['inline_message_id']
+        return message_id, inline_message_id
+
+    def _get_target_chat_ids(self, target):
+        """Validate chat_id targets or return default target (fist defined).
+
+        :param target: optional list of strings or ints (['12234'] or [12234])
+        :return list of chat_id targets (integers)
+        """
+        if target is not None:
+            if isinstance(target, int):
+                if target in self.allowed_chat_ids:
+                    return [target]
+                _LOGGER.warning('BAD TARGET "%s", using default: %s',
+                                target, self._default_user)
+            else:
+                try:
+                    chat_ids = [int(t) for t in target
+                                if int(t) in self.allowed_chat_ids]
+                    if len(chat_ids) > 0:
+                        return chat_ids
+                    _LOGGER.warning('ALL BAD TARGETS: "%s"', target)
+                except (ValueError, TypeError):
+                    _LOGGER.warning('BAD TARGET DATA "%s", using default: %s',
+                                    target, self._default_user)
+        return [self._default_user]
+
+    def _get_msg_kwargs(self, data):
+        """Get parameters in message data kwargs."""
+
+        def _make_row_of_kb(row_keyboard):
+            """Make a list of InlineKeyboardButtons from a list of tuples.
+
+            :param row_keyboard: [(text_b1, data_callback_b1),
+                                  (text_b2, data_callback_b2), ...]
+            """
+            from telegram import InlineKeyboardButton
+            if isinstance(row_keyboard, str):
+                return [InlineKeyboardButton(
+                    key.strip()[1:].upper(),
+                    callback_data=key)
+                        for key in row_keyboard.split(",")]
+            elif isinstance(row_keyboard, list):
+                return [InlineKeyboardButton(
+                    text_btn, callback_data=data_btn)
+                        for text_btn, data_btn in row_keyboard]
+            else:
+                raise ValueError(str(row_keyboard))
+
+        # Defaults
+        params = dict(parse_mode=self._parse_mode,
```

Comment:
I hardly see anyone use dicts like this. It's actually worse perf to create a dict like this compared to specifying it with dict syntax:

```python
params = {
    'parse_mode': self._parse_mode,
    # etc…
}
```